src/main/java/eu/europeana/metis/sandbox/service/workflow/MediaProcessingServiceImpl.java

```diff
@@ -1,13 +1,152 @@
 
+  private static final Logger LOGGER = LoggerFactory.getLogger(MediaProcessingServiceImpl.class);
+
+  private final RdfConverterFactory converterFactory;
+  private final MediaProcessorFactory processorFactory;
+  private final ThumbnailStoreService thumbnailStoreService;
+
+  public MediaProcessingServiceImpl(
+      RdfConverterFactory converterFactory,
+      MediaProcessorFactory processorFactory,
+      ThumbnailStoreService thumbnailStoreService) {
+    this.converterFactory = converterFactory;
+    this.processorFactory = processorFactory;
+    this.thumbnailStoreService = thumbnailStoreService;
+  }
+
   @Override
-  public Record processMedia(Record record) {
-    return record;
+  public RecordInfo processMedia(Record record) {
+    requireNonNull(record, "Record must not be null");
+
+    var inputRdf = record.getContent();
+    var rdfDeserializer = getRdfDeserializer(record);
+
+    // Get resource entries
+    var resourceEntries = getRdfResourceEntries(record, inputRdf, rdfDeserializer);
+
+    var rdfForEnrichment = getEnrichedRdf(record, inputRdf, rdfDeserializer);
+
+    List<RecordError> recordErrors = new LinkedList<>();
+    try (MediaExtractor extractor = processorFactory.createMediaExtractor()) {
+      for (RdfResourceEntry entry : resourceEntries) {
+        processResourceEntry(record, rdfForEnrichment, extractor, entry, recordErrors);
+      }
+    } catch (MediaProcessorException | IOException e) {
+      throw new RecordProcessingException(record.getRecordId(), e);
+    }
+
+    // Get output rdf bytes
+    var rdfSerializer = getRdfSerializer(record);
+    byte[] outputRdf = getOutputRdf(record, rdfSerializer, rdfForEnrichment);
+
+    return new RecordInfo(Record.from(record, outputRdf), recordErrors);
+  }
+
+  private void processResourceEntry(Record record, EnrichedRdf rdfForEnrichment,
+      MediaExtractor extractor,
+      RdfResourceEntry entry, List<RecordError> recordErrors) {
+    try (ResourceExtractionResult extraction = extractor.performMediaExtraction(entry)) {
+      if (nonNull(extraction)) {
+        // Store thumbnails
+        storeThumbnails(record, extraction.getThumbnails(), recordErrors);
+        // Add result to RDF
+        rdfForEnrichment.enrichResource(extraction.getMetadata());
+      }
+    } catch (MediaExtractionException | IOException e) {
+      LOGGER.warn("Error while extracting media for record {}. ", record.getRecordId(), e);
+      // collect warn
+      recordErrors.add(new RecordError(e));
+    }
+  }
+
+  private void storeThumbnails(Record record, List<Thumbnail> thumbnails,
+      List<RecordError> recordErrors) {
+    if (nonNull(thumbnails)) {
+      try {
+        thumbnailStoreService.store(thumbnails);
+      } catch (ThumbnailStoringException e) {
+        LOGGER.warn("Error while storing thumbnail for record {}. ", record.getRecordId(), e);
+        // collect warn
+        recordErrors.add(new RecordError(e));
```

Comment:
Now I am just collecting the errors and logging a warn when an entry has any issue